### PR TITLE
UDP Packet Listener

### PR DESCRIPTION
**Pull request by Kairi700**

### CODE REVIEW

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Created by Lauren Hsu on 3/4/2017.
+ *
+ * UDPListener for receiving the IP addresses of the minibots currently active
+ */
+public class UDPConnectionListener extends Thread{
+
+    /**The amount of seconds needed to pass before evicting an address*/
+    private static int UPDATE_THRESHOLD = 40;
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: not static

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**Contains the current list of IP addresses and when they last contacted the server*/
+    private static Hashtable<String, Timestamp> ipList = new Hashtable<>();
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: not static

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
+package basestation.bot.connection;
+
+import java.io.IOException;
+import java.net.*;
+import java.sql.Timestamp;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Created by Lauren Hsu on 3/4/2017.
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: No author comments :P

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ */
+public class UDPConnectionListener extends Thread{
+
+    /**The amount of seconds needed to pass before evicting an address*/
+    private static int UPDATE_THRESHOLD = 40;
+
+    /**Contains the current list of IP addresses and when they last contacted the server*/
+    private static Hashtable<String, Timestamp> ipList = new Hashtable<>();
+
+    /**Initializes a thread running an instance of UDPConnectionListener**/
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: */ not **/

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    //Taken as a reference from
+    //http://stackoverflow.com/questions/5472269/java-datagramsocket-listening-on-a-broadcast-address
+    @Override
+    public void run (){
+        try {
+            //Receives a packet from any IP address
+            DatagramSocket socket = new DatagramSocket(5001, InetAddress.getByName("0.0.0.0"));
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: Make port a constant: public static final int PORT = 5001;

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            socket.setBroadcast(true);
+            System.out.println("Listen on " + socket.getLocalAddress() + " from " + socket.getInetAddress() + " port " + socket.getBroadcast());
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: Don't print this.

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            byte[] buf = new byte[512];
+            DatagramPacket packet = new DatagramPacket(buf, buf.length);
+
+            while (true) {
+                //Start accepting packets
+                System.out.println("Waiting for data");
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: No print

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                socket.receive(packet);
+                //System.out.println(packet);
+                System.out.println("Data received at: "+getCurrentTime());
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: No print

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            socket.setBroadcast(true);
+            System.out.println("Listen on " + socket.getLocalAddress() + " from " + socket.getInetAddress() + " port " + socket.getBroadcast());
+            byte[] buf = new byte[512];
+            DatagramPacket packet = new DatagramPacket(buf, buf.length);
+
+            while (true) {
+                //Start accepting packets
+                System.out.println("Waiting for data");
+                socket.receive(packet);
+                //System.out.println(packet);
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: Delete comment

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                System.out.println("Data received at: "+getCurrentTime());
+
+                //Get packet header (IP address) as a string for storage
+                String address = packet.getAddress().toString();
+                System.out.println("Sent from: "+address);
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: Delete print

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+
+    /** Returns the set of IP addresses currently being tracked
+     *  by the UDP server
+     *
+     * @returns a set of IP addresses currently being tracked
+     * */
+    //public static Set<String> getAddressList(){return ipList;}
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: Delete comment

File: src/main/java/basestation/bot/connection/UDPConnectionListener.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        return new Timestamp(cal.getTime().getTime());
+    }
+
+
+
+    /** Removes IP addresses from ipList that have not be used in a certain
+     * time frame, designated by UPDATE_THRESHOLD
+     *
+     * */
+    private static void cleanOut(){
```
Comment (TrevorEdwards) [inline src/main/java/basestation/bot/connection/UDPConnectionListener.java]: not static

File: test/UDPTest.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import basestation.BaseStation;
+import basestation.bot.connection.UDPConnectionListener;
+import org.junit.Test;
+
+import java.util.Set;
+
+/**
+ * Created by Administrator on 2/15/2017.
```
Comment (TrevorEdwards) [inline test/UDPTest.java]: No created comment

**Review by TrevorEdwards — CHANGES_REQUESTED** [review]

**Comment by Kairi700** [comment]
Made suggested changes, please check =)

**Review by TrevorEdwards — APPROVED** [review]

**Pull request merged**